server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByEngineDataSet.java

```diff
@@ -200,28 +161,62 @@ public boolean hasNextWithoutConstraint() {
     return true;
   }
 
-  /**
-   * add natural months based on the startTime to avoid edge cases, ie 2/28
-   *
-   * @param startTime current start time
-   * @param numMonths numMonths is updated in hasNextWithoutConstraint()
-   * @return nextStartTime
-   */
-  public static long calcIntervalByMonth(long startTime, long numMonths) {
-    Calendar calendar = Calendar.getInstance();
-    calendar.setTimeZone(SessionManager.getInstance().getCurrSessionTimeZone());
-    calendar.setTimeInMillis(startTime);
-    boolean isLastDayOfMonth =
-        calendar.get(Calendar.DAY_OF_MONTH) == calendar.getActualMaximum(Calendar.DAY_OF_MONTH);
-    calendar.add(Calendar.MONTH, (int) (numMonths));
-    if (isLastDayOfMonth) {
-      calendar.set(Calendar.DAY_OF_MONTH, calendar.getActualMaximum(Calendar.DAY_OF_MONTH));
+  @Override
+  public RowRecord nextWithoutConstraint() throws IOException {
+    if (!hasCachedTimeInterval) {
+      throw new IOException(
+          "need to call hasNext() before calling next()" + " in GroupByWithValueFilterDataSet.");
     }
-    return calendar.getTimeInMillis();
+    hasCachedTimeInterval = false;
+    curAggregateResults = getNextAggregateResult();
+    return constructRowRecord(curAggregateResults);
   }
 
-  @Override
-  public abstract RowRecord nextWithoutConstraint() throws IOException;
+  protected AggregateResult[] getNextAggregateResult() throws IOException {
+    throw new UnsupportedOperationException("Should call exact sub class!");
```

Comment:
Try to make it an abstract method instead of throwing Exception.

Reply:
fixed